Move Dynamo Tutorials (#2120)
pytorch/tutorials: advanced_source/GuardsOverviewPt1.rst

How TorchDynamo Works - Pt1 - From user code to guards
======================================================

> Welcome

Reviewer: Suggesting to delete this header.

-------

> Welcome to the first entry in a series of technical documents outlining

Reviewer: I think, here we need to introduce what TorchDynamo so that the users who know nothing about TorchDynamo get a grasp of it. Example:

TorchDynamo is a Python-level just-in-time (JIT) compiler that to helps you accelerate unmodified PyTorch programs. TorchDynamo supports arbitrary PyTorch code, control flow, mutation, and dynamic shapes. Getting started with TorchDynamo is easy — just add a a single line decorator `torch._dynamo.optimize()`.
```suggestion
```

how TorchDynamo works.

> We assume the reader already has a working knowledge of PyTorch, and

Reviewer: ```suggestion
Before you can run this tutorial, you must have a working knowledge of PyTorch, and
```

Python, and a basic understanding of TorchDynamo’s role in the PyTorch
Ecosystem. The target audience is engineers and researchers looking to
learn more about how TorchDynamo operates under the hood.

Using Dynamo
------------

From a UX perspective, TorchDynamo is very easy to use. The user invokes
``torchdynamo.optimize`` as an annotation:

.. code:: py

   @torchdynamo.optimize(my_compiler)
   def fn_foo(bar):

Where a complete example looks like this:

.. code:: py

   from typing import List
   import torch
   import torchdynamo

   def my_compiler(gm: torch.fx.GraphModule, example_inputs: List[torch.Tensor]):
       print("my_compiler() called with FX graph:")
       gm.graph.print_tabular()
       return gm.forward  # return a python callable

   @torchdynamo.optimize(my_compiler)
   def toy_example(a, b):
       x = a / (torch.abs(a) + 1)
       if b.sum() < 0:
           b = b * -1
       return x * b

   for _ in range(100):
       toy_example(torch.randn(10), torch.randn(10))

This allows TorchDynamo to capture the interpreted Python frames, grab
any and all relevant information, and speed things up wherever it can.

> The speedup comes from a few places, and can be rather dependent on the

Reviewer: ```suggestion
The speedup comes from a few places and can be rather dependent on the
```

> backend (my_compiler above) provided, but the one speedup we care about

Reviewer: ```suggestion
backend (`my_compiler` in the code block above) provided, but the one speedup that is most important
```

> most for today’s overview is **caching**. Caching itself is not a direct

Reviewer: ```suggestion
for this overview is **caching**. Caching itself is not a direct
```

> speedup, so much as a critical enablement to allow us to prevent

Reviewer: ```suggestion
speedup, but a critical enablement that allows us to prevent
```

> recompilation. We dig a hole with dynamo, and caching allows us to get

Reviewer: We need to rephrase "We dig a hole with dynamo, and caching allows us to get out" or remove

> out. Its a speedup from that perspective, but relatively neutral when

Reviewer: ```suggestion
out. It is a speedup from that perspective, but relatively neutral when
```

all things are considered - however, it enables us to hold perf
neutrality while then enabling backends - the true source of our
speedups.

> With even a pass-through no-op backend provided:

Reviewer: This looks like an unfinished thought. And should it be:  "Even with a pass-trhough ..."

.. code:: py

   def my_compiler(gm: torch.fx.GraphModule, example_inputs: List[torch.Tensor]):
       return gm.forward

> We can see TorchDynamo speeding up Python execution quite a bit, even on

Reviewer: ```suggestion
We can see TorchDynamo speeding up Python execution quite a bit, even with
```

regular Python, not just PyTorch.

Caching and Guards Overview
---------------------------

TorchDynamo operates through caching transformed (by TorchDynamo) user
bytecode. When we receive a frame for evaluation, we check if the
**objects referenced in the frame have changed** in certain ways, and if
not, we read the previously transformed user bytecode to evaluate it.

> The details of how we do this will be saved for a later writeup.

Reviewer: Remove: `The details of how we do this will be saved for a later writeup.`
